File: Week 5/LeftFactoring.cpp
#include <iostream>
#include <string>
using namespace std;
int main()
{
	int n,j,l,i,m;
	int len[10] = {};
    string a, b1, b2, flag;
    char c;
    cout << "Enter the Parent Non-Terminal : ";
    cin >> c;
    a.push_back(c);
    b1 += a + "\'->";
    b2 += a + "\'\'->";;
    a += "->";
    cout << "Enter total number of productions : ";
    cin >> n;
    for (i = 0; i < n; i++)
    {
        cout << "Enter the Production " << i + 1 << " : ";
        cin >> flag;
        len[i] = flag.size();
        a += flag;
        if (i != n - 1)
        {
        a += "|";	
		}
    }
    cout << "The Production Rule is : " << a << endl;
    char x = a[3];
    for (i = 0, m = 3; i < n; i++)
    {
        if (x != a[m])
        {
             while (a[m++] != '|');
        }
        else
        {
        	 if (a[m + 1] != '|')
            {
                b1 += "|" + a.substr(m + 1, len[i] - 1);
                a.erase(m - 1, len[i] + 1);
            }
            else
            {
            	b1 += "#";
                a.insert(m + 1, 1, a[0]);
                a.insert(m + 2, 1, '\'');
                m += 4;
            }
        }
    }
    char y = b1[6];
    for (i = 0, m = 6; i < n - 1; i++)
    {
        if (y == b1[m])
        {
            if (b1[m + 1] != '|')
            {
                flag.clear();
                for (int s = m + 1; s < b1.length(); s++)
                {
                     flag.push_back(b1[s]);	
				}
                b2 += "|" + flag;
                b1.erase(m - 1, flag.length() + 2);
            }
            else
            {
            	b1.insert(m + 1, 1, b1[0]);
                b1.insert(m + 2, 2, '\'');
                b2 += "#";
                m += 5;
            }
        }
    }
    b2.erase(b2.size() - 1);
    cout << "After Left Factoring : " << endl;
    cout << a << endl;
    cout << b1 << endl;
    cout << b2 << endl;
    return 0;
}

File: Week 5/Leftfactoring.py
# Code for productions where all terms are in left factoring.
# A->ab|ac|ag|a+c or A->abG|abC|ab+c etc work's perfectly.Non-left factoring productions messes the code.
#A->ab|ac|epsilon won't function
def leftFactoring(s):
    k=[]
    l=[]
    n=""
    k=s.split('->')
    l=k[1].split('|')
    for i in range(0,len(l)-1):  #0 1 2 
       for j in range(0,min(len(l[i]),len(l[i+1]))):  # 0 1 2
           if(l[i][j]==l[i+1][j]):
               if l[i][j] not in n:
                   n=n+l[i][j]

    print(k[0]+'->'+n+"R")
    m=k[1].split(n)
    print("R->",end="")
    for i in range(1,len(m)):
       print(m[i],end="")
    
s=input("Enter the production: ")
while(True):
    checkFactoring(s)
    leftFactoring(s)
    print("\ndo you have another production?")
    T=input("y/n:")
    if T=='y':
        s=input("Enter the production: ")
    elif T=='n':
        break
